src/browser/modules/D3Visualization/lib/visualization/components/Visualization.ts

```diff
@@ -235,129 +132,115 @@ const vizFn = function (
           })`
       )
 
-    for (const renderer of vizRenderers.relationship) {
-      relationshipGroups.call(renderer.onTick, viz)
-    }
+    relationshipRenderer.forEach(renderer =>
+      relationshipGroups.call(renderer.onTick, this)
+    )
   }
 
-  const force = layout.init(render)
+  private updateNodes() {
+    const nodes = this.graph.nodes()
+    this.geometry.onGraphChange(this.graph, {
+      updateNodes: true,
+      updateRelationships: false
+    })
 
-  viz.update = function (options: {
-    updateNodes: boolean
-    updateRelationships: boolean
-    precompute?: boolean
-  }) {
-    if (!graph) {
-      return
-    }
+    const nodeGroups = this.container
+      .select('g.layer.nodes')
+      .selectAll<SVGGElement, VizNode>('g.node')
+      .data(nodes, d => d.id)
+      .join('g')
+      .attr('class', 'node')
+      .call(nodeEventHandlers, this.trigger, this.forceSim.simulation)
+      .classed('selected', node => node.selected)
 
-    geometry.onGraphChange(graph, options)
+    nodeRenderer.forEach(renderer =>
+      nodeGroups.call(renderer.onGraphChange, this)
+    )
 
-    const layers = container
-      .selectAll('g.layer')
-      .data(['relationships', 'nodes'])
-      .join('g')
-      .attr('class', d => `layer ${d}`)
+    for (const renderer of nodeMenuRenderer) {
+      nodeGroups.call(renderer.onGraphChange, this)
```

Comment:
At first I thought you were doing the same as on line 157 again by mistake, but i see now that it's iterating over different renders. Could change this second loop to also be forEach to make is easier to see that it's not duplicated by mistake?

Reply:
Good spot, I missed this :+1: